ibverbs/src/lib.rs

```diff
@@ -533,6 +552,94 @@ impl<'ctx> CompletionQueue<'ctx> {
+    /// alone is insufficient, such as in low-traffic scenarios.
+    ///
+    /// The returned slice reflects completed work requests (e.g., sends, receives) from the
+    /// associated Work Queue. Not all fields in `ibv_wc` are valid unless the status is
+    /// `IBV_WC_SUCCESS`.
+    ///
+    /// # Errors
+    /// - `TimedOut`: If the timeout expires before any completions are available.
+    /// - System errors: From underlying calls like `req_notify_cq`, `poll`, or `ibv_get_cq_event`.
+    ///
+    /// # Notes
+    /// - This method blocks, unlike `poll`, but avoids busy-waiting by leveraging CQ events and
+    ///   system polling.
+    /// - Callers must ensure the CQ does not overrun (exceed its capacity), as this triggers an
+    ///   `IBV_EVENT_CQ_ERR` async event, rendering the CQ unusable.
+    pub fn wait<'c>(
+        &self,
+        completions: &'c mut [ffi::ibv_wc],
+        timeout: Option<Duration>,
+    ) -> io::Result<&'c mut [ffi::ibv_wc]> {
+        let c = completions as *mut [ffi::ibv_wc];
+
+        loop {
+            let completions = self.poll(unsafe { &mut *c })?;
+            if !completions.is_empty() {
+                return Ok(completions);
+            }
+
+            let ctx: *mut ffi::ibv_context = unsafe { *self.cq }.context;
+            let errno = unsafe {
+                let ops = &mut { &mut *ctx }.ops;
+                ops.req_notify_cq.as_mut().unwrap()(self.cq, 0)
+            };
+            if errno != 0 {
+                return Err(io::Error::from_raw_os_error(errno));
+            }
+
+            let completions = self.poll(unsafe { &mut *c })?;
+            if !completions.is_empty() {
+                return Ok(completions);
+            }
+
+            let mut pollfd = libc::pollfd {
+                fd: unsafe { *self.cc }.fd,
+                events: libc::POLLIN,
+                revents: 0,
+            };
+            let rc = unsafe {
+                libc::poll(
+                    &mut pollfd,
+                    1,
+                    timeout.map_or(-1, |x| x.as_millis() as libc::c_int),
+                )
+            };
+            match rc {
+                -1 => return Err(io::Error::last_os_error()),
+                0 => {
+                    return Err(io::Error::new(
+                        io::ErrorKind::TimedOut,
+                        "Timed out during completion queue wait",
+                    ))
+                }
+                1 => {}
+                _ => unreachable!(),
+            }
+
+            let mut out_cq = std::ptr::null_mut();
+            let mut out_cq_context = std::ptr::null_mut();
+            let rc = unsafe { ffi::ibv_get_cq_event(self.cc, &mut out_cq, &mut out_cq_context) };
+            if rc < 0 {
+                let e = io::Error::last_os_error();
+                if e.kind() == io::ErrorKind::WouldBlock {
+                    continue;
+                }
```

Comment:
If `ibv_get_cq_event` supports blocking operations, why do we even need `libc::poll`? Couldn't we just use `ibv_get_cq_event` directly with a blocking socket? Is the only thing preventing us from doing that the lack of a timeout? If so, would be worthwhile to add a comment to that effect.